### ISSUE TITLE
Partial offline mode after restart and RocksDbException

### ISSUE THREAD

**Issue by NoahStahl** [NethermindEth/nethermind#3949]
After running nethermind successfully for a couple months with no issues, I did a graceful restart for maintenance last week and upon starting again, have seen this logged every second for the past week straight:

`Syncing previously downloaded blocks from DB (partial offline mode until it finishes)`

I've restarted nethermind multiple times, same behavior. Some errors are logged suggesting DB corruption -- how can I best resolve this? I've run with `--Sync.FixReceipts true` which did not resolve.

When stopping the process, this error is logged:

```
2022-04-09 07:19:58.9715|ERROR|624|Closing DBs shutdown error. RocksDbSharp.RocksDbException: IO error: ReadFile failed: D:\Nethermind\nethermind_db/mainnet\receipts/038928.sst: Data error (cyclic redundancy check).

   at RocksDbSharp.Native.rocksdb_flush(IntPtr db, IntPtr options)
   at Nethermind.Db.Rocks.DbOnTheRocks.Flush()
   at Nethermind.Db.Rocks.DbOnTheRocks.Dispose(Boolean disposing)
   at Nethermind.Db.Rocks.DbOnTheRocks.Dispose()
   at Nethermind.Db.DbProvider.Dispose()
   at Nethermind.Runner.Ethereum.EthereumRunner.<StopAsync>b__5_8()
   at Nethermind.Runner.Ethereum.EthereumRunner.Stop(Action stopAction, String description)

```

When starting the process, this error is logged:

```
2022-04-09 07:21:20.6166|ERROR|110|BlockchainProcessor encountered an exception. RocksDbSharp.RocksDbException: IO error: ReadFile failed: D:\Nethermind\nethermind_db/mainnet\receipts/038928.sst: Data error (cyclic redundancy check).

   at RocksDbSharp.Native.rocksdb_put(IntPtr db, IntPtr writeOptions, Byte[] key, Int64 keyLength, Byte[] value, Int64 valueLength, ColumnFamilyHandle cf)
   at Nethermind.Db.Rocks.ColumnDb.set_Item(Byte[] key, Byte[] value)
   at Nethermind.Blockchain.Receipts.PersistentReceiptStorage.Insert(Block block, TxReceipt[] txReceipts)
   at Nethermind.Blockchain.Processing.BlockProcessor.StoreTxReceipts(Block block, TxReceipt[] txReceipts)
   at Nethermind.Blockchain.Processing.BlockProcessor.ProcessOne(Block suggestedBlock, ProcessingOptions options, IBlockTracer blockTracer)
   at Nethermind.Blockchain.Processing.BlockProcessor.Process(Keccak newBranchStateRoot, List`1 suggestedBlocks, ProcessingOptions options, IBlockTracer blockTracer)
   at Nethermind.Blockchain.Processing.BlockchainProcessor.ProcessBranch(ProcessingBranch processingBranch, ProcessingOptions options, IBlockTracer tracer)
   at Nethermind.Blockchain.Processing.BlockchainProcessor.Process(Block suggestedBlock, ProcessingOptions options, IBlockTracer tracer)
   at Nethermind.Blockchain.Processing.BlockchainProcessor.RunProcessingLoop()
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
```

**To Reproduce**
Generally, not sure. I suspect the database error means I have corruption.

**Expected behavior**
No errors, expected normal mode.

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Version 1.12.4

**Comment by NoahStahl** [comment]
Just noting this happened again today, upon the first graceful restart to apply updates. I had previously done a full resync and abandoned the previous state to get back to normal function.

It seems like stopping the instance is not safe.

**Comment by asdacap** [comment]
Just checking, according to SMART, how is the health of you drive?

**Comment by NoahStahl** [comment]
There are no signs of issues with the hardware, everything reports healthy. Also, Nethermind runs fine for weeks with all the usual disk activity and no apparent problems. It is just when it is stopped that this issue manifests, which in my experience suggests there's something amiss with handling the process termination? I notice there is already a pretty comprehensive graceful shutdown phase, and this has been invoked each time the problem has occurred. So perhaps there is something more to be done there?

**Comment by begetan** [comment]
It is not an issue, it is a feature according to this ticket: https://github.com/NethermindEth/nethermind/issues/2864

**Comment by MaxTeiger** [comment]
Did you find a way to make the node up & running again ? (I mean re-syncing from the last processed block when it stops)

For the moment I had to setup a new node from scratch, but it would be very convenient for me to make the other one on track again 😄 

Thanks !

**Comment by MaxTeiger** [comment]
> I always had to do a full resync when this happened. Fortunately it hasn't recurred in recent memory.

So supposedly, you would have to recreate a new node every time it crashes?

**Comment by NoahStahl** [comment]
Just occurred an hour ago, node down with sudden `(cyclic redundancy check)` error after running fine for months.

```
2022-10-25 12:51:02.7399|ERROR|513|engine_newPayloadV1 threw an unexpected exception. RocksDbSharp.RocksDbException: IO error: ReadFile failed: D:\Nethermind\nethermind_db/mainnet\blocks/169436.sst: Data error (cyclic redundancy check).

   at RocksDbSharp.Native.rocksdb_put(IntPtr db, IntPtr writeOptions, Byte[] key, Int64 keyLength, Byte[] value, Int64 valueLength, ColumnFamilyHandle cf)
   at Nethermind.Db.Rocks.DbOnTheRocks.set_Item(Byte[] key, Byte[] value)
   at Nethermind.Blockchain.BlockTree.Suggest(Block block, BlockHeader header, BlockTreeSuggestOptions options)
   at Nethermind.Blockchain.BlockTree.SuggestBlockAsync(Block block, BlockTreeSuggestOptions suggestOptions)
   at Nethermind.Merge.Plugin.TimeoutUtils.TimeoutOn[T](Task`1 task, Task timeoutTask)
   at Nethermind.Merge.Plugin.Handlers.V1.NewPayloadV1Handler.ValidateBlockAndProcess(Block block, BlockHeader parent, ProcessingOptions processingOptions)
   at Nethermind.Merge.Plugin.Handlers.V1.NewPayloadV1Handler.HandleAsync(ExecutionPayloadV1 request)
   at Nethermind.Merge.Plugin.EngineRpcModule.engine_newPayloadV1(ExecutionPayloadV1 executionPayload) 
```

**Comment by kamilchodola** [comment]
@NoahStahl @MaxTeiger 
I'm not sure if You are aware but on 1.15.0 we added a logic which will try to restore DB after unexpected issues like hardware issues or power cuts which **MAY** break DB but not necessarily it will.
#4911 

Let me know if You will have this problem again - on my side I'll try my best to reproduce this but not having those problems recently.

**Comment by kamilchodola** [comment]
@NoahStahl @MaxTeiger 
Is any of You had this recently?
I did not reproduced it for long time now and assuming our improvements in that area I assume it can be closed and may be reopened if anyone would spot that again.

**Comment by MaxTeiger** [comment]
Hello @kamilchodola ! 

No I did not face this problem recently :)

**Comment by NoahStahl** [comment]
@kamilchodola Thanks for checking -- I have not seen this happen again since the original reports. I will mark as closed 👍🏽

**Comment by EvilJordan** [comment]
Sorry to re-open and ping all of you, but I thought I'd update as this just happened to me.

I started up my backup node in preparation for the move to `HalfPath` (that's another issue), and Nethermind had a few months of catching up to do.

It finally finished, but then started spitting out errors like this:

```
Error when handling Id:609866, eth_getLogs([{"topics":["0x649bbc62d0e31342afea4e5cd82d4049e7e1ee912fc0889aa790803be39038c5"],"fromBlock":"0x12d169e","toBlock":"0x12d169e","address":["0x00000000219ab540356cBB839Cbe05303d7705Fa"]}]) | { 
  "jsonrpc": "2.0", 
  "error": { 
    "code": -32001, 
    "message": "Receipt not available for From block 19732126." 
  }, 
  "id": 609866 
} 
```

I restarted to see if that would fix, and those errors went away and now I see this:
`Syncing previously downloaded blocks from DB (partial offline mode until it finishes)`

There's no indication of how far along this is or how long it will take.

**Comment by kamilchodola** [comment]
@EvilJordan 
This message (partial offline mode) appears when you downloaded all blocks missing but not yet processed them.
Downloading of data is faster than processing of it so it is quite common behavior on very old backups after restart.

Seems that missing receipt is pretty close to the head so it may mean that CL waits for receipts of that block but we do not have them yet as this block was not processed (but probably already downloaded).

Hope it clarifies a bit.
